File: dotnet/ProducerTask.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Confluent.Kafka;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace KafkaTool;


public static class ProducerTask
{
    private static readonly ILogger Log = LoggerFactory
        .Create(builder => builder.AddSimpleConsole(options =>
        {
            options.SingleLine = true;
            options.TimestampFormat = "HH:mm:ss ";
        }))
        .CreateLogger("Log");

    public static Task GetTask(ProducerConsumerSettings settings, ProducerConsumerData data, int producerIndex)
    {
        return Task.Run(async () =>
        {
            var cancellationTokenSource = new CancellationTokenSource();
            IConfiguration configuration = new ConfigurationBuilder()
                .AddInMemoryCollection(settings.ConfigDictionary)
                .Build();

            var logger = LoggerFactory
                .Create(builder => builder.AddSimpleConsole(options =>
                {
                    options.SingleLine = true;
                    options.TimestampFormat = "HH:mm:ss ";
                }))
                .CreateLogger($"Producer{producerIndex}:");

            logger.Log(LogLevel.Information, "Starting producer task:");

            Exception e = null;
            var producerConfig = new ProducerConfig(settings.ConfigDictionary);

            var producer = new ProducerBuilder<long, long>(
                    producerConfig.AsEnumerable().Concat(configuration.AsEnumerable()))
                .SetLogHandler(
                    (a, b) =>
                    {
                        if (!b.Message.Contains(": Disconnected (after ",
                                StringComparison.OrdinalIgnoreCase))
                        {
                            logger.LogInformation($"kafka-log Facility:{b.Facility}, Message{b.Message}");
                        }
                    })
                .SetErrorHandler((_, e) => logger.Log(LogLevel.Error,
                    $"Producer error: reason={e.Reason}, IsLocal={e.IsLocalError}, IsBroker={e.IsBrokerError}, IsFatal={e.IsFatal}, IsCode={e.Code}"))
                .Build();

            var oneMillisecond = TimeSpan.FromMilliseconds(1);
            var swTimestamp = Stopwatch.GetTimestamp();
            double messagesToSend = 0;
            double messagesRate = (double)settings.MessagesPerSecond / 1000;
            var burstCycleSw = Stopwatch.StartNew();
            var resetMessagesToSend = false;
            if (settings.Topics % settings.Producers != 0)
            {
                throw new Exception($"Cannot evenly schedule {settings.Topics} on a {settings.Producers} producers!");
            }

            var topicsPerProducer = settings.Topics / settings.Producers;
            var topicPartitions = new List<TopicPartition>(
                from topic in Enumerable.Range(0, settings.Topics)
                from partition in Enumerable.Range(0, settings.Partitions)
                select new TopicPartition(Utils.GetTopicName(settings.TopicStem, topic), new Partition(partition)));

            for (var currentValue = 0L;; currentValue++)
            for (var topicIndex = 0; topicIndex < topicsPerProducer; topicIndex++)
            {
                for (var partition = 0; partition < settings.Partitions; partition++)
                {
                    var topicPartition = topicPartitions[(topicIndex + producerIndex * topicsPerProducer) * settings.Partitions + partition];
                    if (e != null)
                    {
                        throw e;
                    }

                    do
                    {
                        if (settings.BurstCycle > 0)
                        {
                            if (burstCycleSw.ElapsedMilliseconds > settings.BurstCycle)
                            {
                                burstCycleSw = Stopwatch.StartNew();
                                resetMessagesToSend = true;
                            }

                            if (burstCycleSw.ElapsedMilliseconds < settings.BurstDuration)
                            {
                                messagesRate = (double)settings.BurstMessagesPerSecond / 1000;
                            }
                            else
                            {
                                if (resetMessagesToSend)
                                {
                                    messagesToSend = 0;
                                    resetMessagesToSend = false;
                                }

                                messagesRate = (double)settings.MessagesPerSecond / 1000;
                            }
                        }

                        var currentTimestamp = Stopwatch.GetTimestamp();
                        var elapsed = Stopwatch.GetElapsedTime(swTimestamp, currentTimestamp);
                        if (elapsed > oneMillisecond)
                        {
                            swTimestamp = currentTimestamp;
                            messagesToSend += messagesRate * elapsed.TotalMilliseconds;
                        }

                        if (messagesToSend < 1)
                        {
                            await Task.Delay(TimeSpan.FromMilliseconds(1));
                        }
                    } while (messagesToSend < 1);
 
                    messagesToSend -= 1.0;

                    var msg = new Message<long, long> { Key = partition, Value = currentValue, Timestamp = new Timestamp(DateTime.UtcNow)};
                    
                    bool produced = false;
                    do
                    {
                        try
                        {
                            producer.Produce(topicPartition, msg,
                                (deliveryReport) =>
                                {
                                    if (deliveryReport.Error.Code != ErrorCode.NoError)
                                    {
                                        using var adminClient = Utils.GetAdminClient(settings.ConfigDictionary);
                                        var topicMetadata = adminClient.GetMetadata(deliveryReport.Topic,
                                            TimeSpan.FromSeconds(30));
                                        var partitionsCount = topicMetadata.Topics.Single().Partitions.Count;

                                        if (e == null)
                                        {
                                            e = new Exception(
                                                $"DeliveryReport.Error, Code = {deliveryReport.Error.Code}, Reason = {deliveryReport.Error.Reason}" +
                                                $", IsFatal = {deliveryReport.Error.IsFatal}, IsError = {deliveryReport.Error.IsError}" +
                                                $", IsLocalError = {deliveryReport.Error.IsLocalError}, IsBrokerError = {deliveryReport.Error.IsBrokerError}" +
                                                $", topic = {deliveryReport.Topic}, partition = {deliveryReport.Partition.Value}, partitionsCount = {partitionsCount}");
                                        }
                                    }
                                    else
                                    {
                                        data.IncrementProduced();
                                        var latency = DateTime.UtcNow - deliveryReport.Message.Timestamp.UtcDateTime;
                                        data.DigestProducerLatency(latency.TotalSeconds);
                                    }
                                });
                            produced = true;
                        }
                        catch (ProduceException<long, long> exception)
                        {
                            if (exception.Error.Code != ErrorCode.Local_QueueFull)
                            {
                                logger.LogInformation($"Handling exception");
                                throw;
                            }

                            await Task.Delay(TimeSpan.FromMilliseconds(1), cancellationTokenSource.Token);
                        }
                    } while (!produced);
                }
            }
        });
    }
}

File: python/src/kafka_tool/producer_task.py
import logging
import time
import threading
from typing import Dict

from confluent_kafka import Producer, KafkaError

from .data import ProducerConsumerData
from .settings import ProducerConsumerSettings
from .utils import get_admin_client, get_topic_name
from threading import Thread
import functools
from .polling_producer import PollingProducer

log = logging.getLogger(__name__)

def run_producer_task(
        config: Dict[str, str],
        settings: ProducerConsumerSettings,
        data: ProducerConsumerData,
        producer_index: int,
        shutdown: threading.Event):

    if settings.topics % settings.producers != 0:
        raise Exception(
            f"Cannot evenly schedule {settings.topics} topics on {settings.producers} producers")

    topics_per_producer = settings.topics // settings.producers

    producer = PollingProducer(config)
    log.info("Running producer task %d", producer_index)

    exception = None

    def delivery_report(err, msg):
        nonlocal exception

        if err is not None:
            log.error(f"Delivery failed for {msg.topic()}: {err}")
            if exception is None:
                admin_client = get_admin_client(config)
                topic_metadata = admin_client.list_topics(topic=msg.topic(), timeout=30)
                partitions_count = len(topic_metadata.topics[msg.topic()].partitions)
                exception = Exception(
                    f"DeliveryReport.Error, Code = {err.code()}, Reason = {err.str()}"
                    f", IsFatal = {err.fatal()}, IsError = {err.code() != KafkaError.NO_ERROR}"
                    f", IsLocalError = {err.retriable()}"
                    f", topic = {msg.topic()}, partition = {msg.partition()}, partitionsCount = {partitions_count}"
                )
        else:
            data.increment_produced()

    messages_until_sleep = 0
    time_since_sleep = time.monotonic()

    current_value = 0
    while not shutdown.is_set():
        for topic_index in range(topics_per_producer):
            topic_name = get_topic_name(
                settings.topic_stem, topic_index + producer_index * topics_per_producer)
            for k in range(settings.partitions * 7):
                if exception is not None:
                    raise exception

                if messages_until_sleep <= 0:
                    elapsed = time.monotonic() - time_since_sleep
                    if elapsed < 0.1:
                        time.sleep(0.1 - elapsed)
                    time_since_sleep = time.monotonic()
                    messages_until_sleep = settings.messages_per_second // 10

                messages_until_sleep -= 1

                while True:
                    try:
                        producer.produce(topic_name, key=str(k), value=str(current_value), on_delivery=delivery_report)
                        break
                    except BufferError as e:
                        producer.flush()

        current_value += 1